update_and_signal_handlers/atomic_message_handlers.py

```diff
@@ -0,0 +1,179 @@
+import asyncio
+from datetime import timedelta
+import logging
+from typing import Dict, List, Optional
+import uuid
+
+from temporalio import activity, common, workflow
+from temporalio.client import Client, WorkflowHandle
+from temporalio.worker import Worker
+
+# This samples shows off the key concurrent programming primitives for Workflows, especially 
+# useful for workflows that handle signals and updates.
+
+#   - Makes signal and update handlers only operate when the workflow is within a certain state 
+#     (here between cluster_started and cluster_shutdown) using workflow.wait_condition.
+#   - Signal and update handlers can block and their actions can be interleaved with one another and with the main workflow.
+#     Here, we use a lock to protect shared state from interleaved access.
+#   - Running start_workflow with an initializer signal that you want to run before anything else.
+#
+@activity.defn
+async def allocate_nodes_to_job(nodes: List[int], task_name: str) -> List[int]:
+    print(f"Assigning nodes {nodes} to job {task_name}")
+    await asyncio.sleep(0.1)
+
+@activity.defn
+async def deallocate_nodes_for_job(nodes: List[int], task_name: str) -> List[int]:
+    print(f"Deallocating nodes {nodes} from job {task_name}")
+    await asyncio.sleep(0.1)
+
+@activity.defn
+async def find_bad_nodes(nodes: List[int]) -> List[int]:
+    await asyncio.sleep(0.1)
+    bad_nodes = [n for n in nodes if n % 5 == 0]
+    if bad_nodes:
+        print(f"Found bad nodes: {bad_nodes}")
+    return bad_nodes
+
+# ClusterManager keeps track of the allocations of a cluster of nodes. 
+# Via signals, the cluster can be started and shutdown.
+# Via updates, clients can also assign jobs to nodes and delete jobs.
+# These updates must run atomically.
+@workflow.defn
+class ClusterManager:
+    def __init__(self) -> None:
+        self.cluster_started = False
+        self.cluster_shutdown = False
+        # Protects workflow state from interleaved access
+        self.nodes_lock = asyncio.Lock()
+        self.max_assigned_nodes = 0
+
+    @workflow.signal
+    async def start_cluster(self):
+        self.cluster_started = True
+        self.nodes : Dict[Optional[str]] = dict([(k, None) for k in range(25)])
+        workflow.logger.info("Cluster started")
+
+    @workflow.signal
+    async def shutdown_cluster(self):
+        await workflow.wait_condition(lambda: self.cluster_started)
+        self.cluster_shutdown = True
+        workflow.logger.info("Cluster shut down")
+
+    @workflow.update
+    async def allocate_n_nodes_to_job(self, task_name: str, num_nodes: int, ) -> List[int]:
+        await workflow.wait_condition(lambda: self.cluster_started)
+        assert not self.cluster_shutdown
```

Comment:
This would fail the workflow task causing the workflow to hang. Is this intentional? If not, consider using an exception that would fail the update.